### ISSUE TITLE
Add bash completion script for caffe tool

### ISSUE THREAD

**Issue by CDLuminate** [BVLC/caffe#3149]
For Debian systems the usage of this script `caffe` is installing it under directory `/usr/share/bash-completion/completions/`.  I'm not sure where would you like to put this file, so I did not prepare any Pull Request. Thanks!

This is the script:

``` shell
# bash completion for Caffe                              -*- shell-script -*-
# (C) 2015 Lumin <cdluminate@gmail.com>
# License of this script is BSD-2-Clause

_caffe()
{
  local cur prev words cword
  _init_completion -s || return

  local prototxts='@(prototxt)'
  local caffemodels='@(caffemodel)'
  local solverstates='@(solverstate)'
  local caffefiles='@(prototxt|caffemodel|solverstate)'

  local flags='-gpu -iterations -model -snapshot -solver -weights -sighup_effect -sigint_effect'

  if [[ $cword -eq 1 ]]; then
    COMPREPLY=( $( compgen -W 'train test time device_query' -- "$cur" ) )
    return 0
  fi

  if [[ $cword -eq 2 ]]; then
    case ${words[1]} in
    train|test|device_query|time)
      COMPREPLY=( $( compgen -W "$flags" -- "$cur") )
      return 0
      ;;
    *)
      return 0
      ;;
    esac
  fi

  case $prev in
  -gpu|-iterations|-version)
    return 0
    ;;
  -solver|-model)
    _filedir $prototxts
    return 0
    ;;
  -weights)
    _filedir $caffemodels
    return 0
    ;;
  -snapshot)
    _filedir $solverstates
    return 0
    ;;
  -sighup_effect|-sigint_effect)
    COMPREPLY=( $( compgen -W 'snapshot stop none' -- "$cur") )
    ;;
  *)
    COMPREPLY=( $( compgen -W "$flags" -- "$cur") )
    return 0
    ;;
  esac

  # file completion on relevant files
  _filedir "$caffefiles"

  return 0
}
complete -F _caffe caffe

# vim
```

This is base64 code of this script

```
IyBiYXNoIGNvbXBsZXRpb24gZm9yIENhZmZlICAgICAgICAgICAgICAgICAgICAgICAgICAgICAg
LSotIHNoZWxsLXNjcmlwdCAtKi0KIyAoQykgMjAxNSBMdW1pbiA8Y2RsdW1pbmF0ZUBnbWFpbC5j
b20+CiMgTGljZW5zZSBvZiB0aGlzIHNjcmlwdCBpcyBCU0QtMi1DbGF1c2UKCl9jYWZmZSgpCnsK
ICBsb2NhbCBjdXIgcHJldiB3b3JkcyBjd29yZAogIF9pbml0X2NvbXBsZXRpb24gLXMgfHwgcmV0
dXJuCgogIGxvY2FsIHByb3RvdHh0cz0nQChwcm90b3R4dCknCiAgbG9jYWwgY2FmZmVtb2RlbHM9
J0AoY2FmZmVtb2RlbCknCiAgbG9jYWwgc29sdmVyc3RhdGVzPSdAKHNvbHZlcnN0YXRlKScKICBs
b2NhbCBjYWZmZWZpbGVzPSdAKHByb3RvdHh0fGNhZmZlbW9kZWx8c29sdmVyc3RhdGUpJwoKICBs
b2NhbCBmbGFncz0nLWdwdSAtaXRlcmF0aW9ucyAtbW9kZWwgLXNuYXBzaG90IC1zb2x2ZXIgLXdl
aWdodHMgLXNpZ2h1cF9lZmZlY3QgLXNpZ2ludF9lZmZlY3QnCiAgCiAgaWYgW1sgJGN3b3JkIC1l
cSAxIF1dOyB0aGVuCiAgICBDT01QUkVQTFk9KCAkKCBjb21wZ2VuIC1XICd0cmFpbiB0ZXN0IHRp
bWUgZGV2aWNlX3F1ZXJ5JyAtLSAiJGN1ciIgKSApCiAgICByZXR1cm4gMAogIGZpCiAgCiAgaWYg
W1sgJGN3b3JkIC1lcSAyIF1dOyB0aGVuCiAgICBjYXNlICR7d29yZHNbMV19IGluCiAgICB0cmFp
bnx0ZXN0fGRldmljZV9xdWVyeXx0aW1lKQogICAgICBDT01QUkVQTFk9KCAkKCBjb21wZ2VuIC1X
ICIkZmxhZ3MiIC0tICIkY3VyIikgKQogICAgICByZXR1cm4gMAogICAgICA7OwogICAgKikKICAg
ICAgcmV0dXJuIDAKICAgICAgOzsKICAgIGVzYWMKICBmaQoKICBjYXNlICRwcmV2IGluCiAgLWdw
dXwtaXRlcmF0aW9uc3wtdmVyc2lvbikKICAgIHJldHVybiAwCiAgICA7OwogIC1zb2x2ZXJ8LW1v
ZGVsKQogICAgX2ZpbGVkaXIgJHByb3RvdHh0cwogICAgcmV0dXJuIDAKICAgIDs7CiAgLXdlaWdo
dHMpCiAgICBfZmlsZWRpciAkY2FmZmVtb2RlbHMKICAgIHJldHVybiAwCiAgICA7OwogIC1zbmFw
c2hvdCkKICAgIF9maWxlZGlyICRzb2x2ZXJzdGF0ZXMKICAgIHJldHVybiAwCiAgICA7OwogIC1z
aWdodXBfZWZmZWN0fC1zaWdpbnRfZWZmZWN0KQogICAgQ09NUFJFUExZPSggJCggY29tcGdlbiAt
VyAnc25hcHNob3Qgc3RvcCBub25lJyAtLSAiJGN1ciIpICkKICAgIDs7CiAgKikKICAgIENPTVBS
RVBMWT0oICQoIGNvbXBnZW4gLVcgIiRmbGFncyIgLS0gIiRjdXIiKSApCiAgICByZXR1cm4gMAog
ICAgOzsKICBlc2FjCgogICMgZmlsZSBjb21wbGV0aW9uIG9uIHJlbGV2YW50IGZpbGVzCiAgX2Zp
bGVkaXIgIiRjYWZmZWZpbGVzIgoKICByZXR1cm4gMAp9CmNvbXBsZXRlIC1GIF9jYWZmZSBjYWZm
ZQoKIyB2aW0K
```

**Comment by CDLuminate** [comment]
Oops, fixed a little bug in the script
```diff
--- caffe.orig	2015-10-05 09:49:57.326814153 +0000
+++ caffe	2015-10-05 09:49:36.167229171 +0000
@@ -49,6 +49,7 @@
     ;;
   -sighup_effect|-sigint_effect)
     COMPREPLY=( $( compgen -W 'snapshot stop none' -- "$cur") )
+    return 0
     ;;
   *)
     COMPREPLY=( $( compgen -W "$flags" -- "$cur") )
```

**Comment by lukeyeager** [comment]
Works for me. Neat!

**Comment by longjon** [comment]
Cool! Why not PR this as `tools/extra/completions/caffe`, and we can work out installation from there?

**Comment by CDLuminate** [comment]
I'm adding this script in caffe's packaging repo. Will create PR later.

**Comment by CDLuminate** [comment]
Hi, I just created a PR for this script.

https://github.com/BVLC/caffe/pull/4576

**Comment by shelhamer** [comment]
Addressed by #4576
